drgn_tools/lock.py

```diff
@@ -25,18 +25,79 @@
 trapping these two function is sufficient to check the semaphore waiters.
 """
 import argparse
+from collections import defaultdict
 from typing import Set
 
 import drgn
+from drgn import Object
 from drgn import Program
+from drgn.helpers.linux.cpumask import for_each_present_cpu
+from drgn.helpers.linux.percpu import per_cpu
 
 from drgn_tools.bt import bt
 from drgn_tools.bt import bt_has
 from drgn_tools.corelens import CorelensModule
 from drgn_tools.locking import for_each_mutex_waiter
 from drgn_tools.locking import for_each_rwsem_waiter
+from drgn_tools.locking import get_rwsem_info
 from drgn_tools.locking import mutex_owner
 from drgn_tools.locking import show_lock_waiter
+from drgn_tools.locking import tail_osq_node_to_spinners
+from drgn_tools.task import get_current_run_time
+from drgn_tools.task import nanosecs_to_secs
+
+
+def scan_osq_node(prog: Program, verbosity=0) -> None:
+    """
+    Show CPUs spinning to grab sleeping lock(s).
+
+    :param prog: drgn.Program
+    :param verbosity: specify verbosity of report as follows:
+                      0: Show which CPUs are spinning
+                      1: Show which CPUs are spinning and for how long
+                      2: Show spinning CPUs, their spin duration and call stack
+                         till the point of spin
+    """
+
+    osq_spinners: defaultdict = defaultdict(list)
+    for cpu in for_each_present_cpu(prog):
+        osq_node = per_cpu(prog["osq_node"], cpu)
+        if not osq_node.next.value_():
+            continue
+
+        while osq_node.next.value_():
+            osq_node = Object(
+                prog,
+                "struct optimistic_spin_node",
+                address=osq_node.next.value_(),
+            )
+
+        if osq_node.address_ in osq_spinners.keys():
+            continue
+
+        for spinner_cpu in tail_osq_node_to_spinners(osq_node):
+            osq_spinners[osq_node.address_].append(spinner_cpu)
+
+    if not len(osq_spinners):
+        print("There are no spinners on any osq_lock")
+    else:
+        print("There are spinners on one or more osq_lock")
+        for key in osq_spinners.keys():
+            print(f"CPU(s): {osq_spinners[key]} are spinning on same osq_lock")
+            if verbosity >= 1:
+                cpu_list = osq_spinners[key]
+                for cpu in cpu_list:
+                    run_time_ns = get_current_run_time(prog, cpu)
+                    run_time_s = nanosecs_to_secs(run_time_ns)
+                    print(
+                        f"CPU: {cpu} has been spinning for {run_time_s} seconds."
+                    )
```

Comment:
Actually these are not spinlocks. And even for spinlocks as long as someone is spinning , he is spinning with preemption disabled so he should not move away from the CPU . So if the spinners spins for 100 ms and does that 20 times, the above snippet will show only its spin time since it last arrival on CPU. It will not show cumulative spin time. But I get your point about the message. It has scope of being misconstrued :) .How about "has been spinning for X secs, since it last got on CPU" is more appropriate than "not scheduled" since spinner is already on CPU.
These osq spinners are also spinning on a per-cpu variable , similar logic applies to them as well.

Reply:
I started this review thinking of OSQ spinners as the spinlocks and wrote this comment with that in mind, but I realized that was wrong and never updated it... sorry!  (I guess it's the qspinlocks whose code looks very similar to the OSQ spinners. 😫 )

In any case I also worded my complaint badly and it was a bit pedantic... your suggestion looks great.

Reply:
No probs @brenns10, indeed qspinlock code shares similarity and its easy to mistake one for other.I have my spinlock waiters helper based on similar logic, but recently with a UEK-5 bug I found it was not showing up all waiters. Will create a PR for that separately. Thanks for the reviews so far :)